### ISSUE TITLE
:show directive overridden by any CSS class setting display property

### ISSUE THREAD

**Issue by paulofabiano** [surface-ui/surface#394]
Hey, 👋 

I have a component where I use TailwindCSS. One of the classes I use on it is the **flex** one, which defines a **display: flex** for the element.

I needed to implement the **:show** directive to represent the state of my component but even if **:show={{ false }}**, the component is still visible because the **display: flex** from TailwindCSS.

The show directive adds the **hidden** attribute to the element and I do believe this is a more elegant solution but I believe the :show directive must have priority for showing or not the UI, independently about how this is designed.

So, could we consider using **style attribute** for using **display: none** in case the directive is false?

Best,
Paulo

**Comment by Malian** [comment]
@paulofabiano the `hidden` attribute replaced the inline style in #244. If you want to have more info about why we did that change, please look at #243.

According to https://css-tricks.com/the-hidden-attribute-is-visibly-weak/, the `hidden` attribute is weak, so it means that it will be overwritten by CSS class or style, this is the expected behaviour. 

Adding this snippet in your css file should solve your issue:
```css
[hidden] { display: none !important; }
``` 

Phoenix adopted the same strategy: https://github.com/phoenixframework/phoenix_live_reload/issues/109

That being said, I agree that it is counterintuitive. We could probably improve the documentation about the `:show` directive to say that under the hood, it sets the `hidden` attribute. 

@maennchen  as you are the one who worked on the PR that introduces the `hidden` attribute, what do you think we could do to improve UX?

**Comment by paulofabiano** [comment]
Hi, @Malian 
Thanks for your reply on this. Well, I understand the whole approach and I guess I can work with this snippet to quickly solve the issue on my side.
Also I agree some improvements in the documentation would be interesting.

**Comment by maennchen** [comment]
@Malian I'm not sure.

The `hidden` attribute is the only way I know that hides an element while not creating problems with CSP / requiring to define classes in CSS.

I think it is a good default since it just works, the docs should mention it though.

While I prefer to fix it by overriding the hidden attribute with `important`, we could also allow configuring a `class` / `style` as live reload does.

**Comment by Malian** [comment]
@maennchen I like the idea to allow configuring a class but I do not like that Surface does not work out of the box and to make it work as expected, user should define a class and/or a css style.

@paulofabiano I got deeper into the `:show` directive code and  normally the code removes all style and class attributes, so I do not understand why it does not work with your code. Could you please share it with me so I would try to find a solution ?

**Comment by maennchen** [comment]
@Malian i believe the hidden attribute is the most standardized solution since it does work with current browsers out of the box and also doesn't clash with CSP.

We could both set the hidden flag and also inline styles and have a configuration to disable the inline styles. (+ a comment in the docs explaining it for users with CSP)

**Comment by msaraiva** [comment]
After reading this discussion I think we should revert this change and update the style, after all the `:show` directive can only be used in regular HTML tags, not in components. That means you can already use `hidden` directly. By updating the style we at least offer a different solution with a different behaviour.

**Comment by maennchen** [comment]
@msaraiva In my opinion, inline styles are not the way to go or it should at least be able to be overwritten. There's a lot of cases where CSPs apply and there should be a way to override this.

**Comment by Malian** [comment]
@msaraiva VueJS has the same "issue" with the `v-show` directive but today, VueJS still uses `display: none` instead of the `hidden` attribute. Look at https://github.com/vuejs/vue/issues/11864 for more info.

To be honest, I never have used the `:show` directive in my projects. I do not see the real value of this directive in Surface. From my point of view it should be handle by the UI instead of the server. The user could always do: `{...hidden: @condition}` or  `class={my_hidden_class: @condition} `

@maennchen @paulofabiano I am curious about your use cases, could you tell me more?

**Comment by maennchen** [comment]
@Malian I'm only using it for stepped forms where I still want the browser to send the hidden fields.

For everything else, I use an if instead.

**Comment by maennchen** [comment]
Example: https://github.com/jshmrtn/hygeia/blob/26d7a1909132897cab30678b7027c01b1bb7f038/apps/hygeia_web/lib/hygeia_web/live/possible_index_submission_live/create.sface#L46

**Comment by maennchen** [comment]
@Malian I agree that this can be achieved also by setting hidden diretly. But with that reasoning, I would argue that surface should not support the show directive at all and instead tell the user to use dynamic classes or hidden.
